bootstrap/handlers/httpserver.go

```diff
@@ -109,6 +109,15 @@ func (b *HttpServer) BootstrapHandler(
 		return false
 	}
 
+	// Use the common middlewares
+	secretProvider := container.SecretProviderExtFrom(dic.Get)
+	authenticationHook := AutoConfigAuthenticationFunc(secretProvider, lc)
+
+	b.router.Use(authenticationHook)
```

Comment:
Understood the case for `AddCustomRoute(.... Unauthenticated ...)` that you added for `AddRoute`. 

Do you think if we still need to keep the flexibility for `authenticationHook` to define on the route level instead of on the server(router) level as in this PR's new behavior?

Reply:
To be honest, I don't like the way that `auth_middleware.go`  has been modified.

Presume the user wants an unauthenticated route or wants to authenticate in a manner different than we have chosen.  Given the manner in which the solution is implemented today, in order to add additional unauthenticated routes, the user would have to go in and fork our core logic.  This is very much against agile programming conventions, where the preferred mechanism of extending the system is to add code, and having to modify existing code to add functionality is viewed as a failure of the software architecture.

On the other hand, in order to implement OpenZiti zero trust, we will have to put /api/v3/ping on its own port, because Consul, Kubernetes, etc can't talk natively on the zero trust overlay, and we still need to have service health checking.  In this scenario, the design is cleaner if all the microservice API's are authenticated by the middleware, without giving the user the ability to make a "mistake" by implementing an unauthenticated route handler.

I know I am arguing both ways, but my subjective opinion is that putting a very specific "if" statement in the core logic is a worse crime than having to include the authentication middleware specifically.

Reply:
> Long term, I think we would have a simpler design if healthcheck is moved to its own http server,

That would require it to have a separate port, which I am not wild about.

Reply:
> > Long term, I think we would have a simpler design if healthcheck is moved to its own http server,
> 
> That would require it to have a separate port, which I am not wild about.

The alternatives are (1) no Ziti when running in k8s, (2) zitifiy consul [unlikely], (3) give up consul health checks, (4) run a ziti tunneler (create a gigantic hole in the zero trust overlay).  We will have to decide this when we get the openziti ADR written.  But since it is relevant to the discussion and future directions, I thought it was worthy to bring it up, as our decisions here will affect future effort.

Reply:
@bnevis-i Thanks for the addressing all your concerns regarding moving the `authenticationHook`  to the server router level and they all make lots of sense. After discussing with @cloudxxx8 , we'd like to avoid the breaking change and also fit the requirement for users to keep some custom route without authentication, so `authenticationHook` change will be reverted back to define on the API route level instead of on the server router level.

May I know your thought on this? Thanks.

Reply:
@lindseysimple I agree that putting `authenticationHook` back on the routes and keeping `AddCustomRoute` will be the best way to effect the relocation of the core routes as a non-breaking change, where non-breaking is defined by "adopters that have existing routes will not suddenly have authentication applied to them (routes, no the adopter :-))."